watcher/src/watcher_db.rs

```diff
@@ -0,0 +1,252 @@
+// Copyright (c) 2018-2020 MobileCoin Inc.
+
+//! The watcher database
+
+use crate::error::WatcherDBError;
+
+use mc_common::logger::{log, Logger};
+use mc_transaction_core::BlockSignature;
+use mc_util_serial::{decode, encode};
+
+use lmdb::{Cursor, Database, DatabaseFlags, Environment, Transaction, WriteFlags};
+use std::{path::PathBuf, sync::Arc};
+
+/// LMDB Constant.
+const MAX_LMDB_FILE_SIZE: usize = 1_099_511_627_776; // 1 TB
+
+/// Keys used by the `counts` database.
+const NUM_BLOCKS_KEY: &str = "num_blocks";
+
+/// Counts database name.
+pub const COUNTS_DB_NAME: &str = "watcher_db:counts";
+
+/// Block signatures database name.
+pub const BLOCK_SIGNATURES_DB_NAME: &str = "watcher_db:block_signatures";
+
+#[derive(Clone)]
+/// DB for Watcher Node.
+pub struct WatcherDB {
+    /// LMDB Environment (database).
+    env: Arc<Environment>,
+
+    /// Signature store.
+    block_signatures: Database,
+
+    /// Aggregate counts about the watcher db.
+    /// * `NUM_BLOCKS_KEY` ---> number of blocks in the watcher db.
+    counts: Database,
+
+    /// Logger.
+    logger: Logger,
+}
+
+impl WatcherDB {
+    /// Open an existing WatcherDB.
+    pub fn open(path: PathBuf, logger: Logger) -> Result<Self, WatcherDBError> {
+        let env = Arc::new(
+            Environment::new()
+                .set_max_dbs(10)
+                .set_map_size(MAX_LMDB_FILE_SIZE)
+                .open(path.as_ref())?,
+        );
+        let counts = env.open_db(Some(COUNTS_DB_NAME))?;
+        let block_signatures = env.open_db(Some(BLOCK_SIGNATURES_DB_NAME))?;
+
+        Ok(WatcherDB {
+            env,
+            block_signatures,
+            counts,
+            logger,
+        })
+    }
+
+    /// Create a fresh WatcherDB.
+    pub fn create(path: PathBuf) -> Result<(), WatcherDBError> {
+        let env = Arc::new(
+            Environment::new()
+                .set_max_dbs(10)
+                .set_map_size(MAX_LMDB_FILE_SIZE)
+                .open(path.as_ref())?,
+        );
+
+        let counts = env.create_db(Some(COUNTS_DB_NAME), DatabaseFlags::empty())?;
+        env.create_db(
+            Some(BLOCK_SIGNATURES_DB_NAME),
+            DatabaseFlags::DUP_SORT | DatabaseFlags::DUP_FIXED,
+        )?;
+
+        let mut db_txn = env.begin_rw_txn()?;
+        db_txn.put(counts, &NUM_BLOCKS_KEY, &encode(&0), WriteFlags::empty())?;
+        db_txn.commit()?;
+        Ok(())
+    }
+
+    /// Add a signature for a block.
+    pub fn add_block_signatures(
+        &self,
+        block_index: u64,
+        signatures: &[BlockSignature],
+    ) -> Result<(), WatcherDBError> {
+        // NOTE: num_blocks is actually "block height" - some blocks may be missing.
+        // This allows us to skip the origin block, and assumes we are adding signatures
+        // to monotonically increasing block indices.
+        let mut db_txn = self.env.begin_rw_txn()?;
+        let num_blocks = self.num_blocks()?;
```

Comment:
This is not using transactions correctly. Instead of using `db_txn` created above, it creates another transaction.